senders/integration_test.go

```diff
@@ -46,3 +48,53 @@ func TestTLSEndToEnd(t *testing.T) {
 	assert.Equal(t, 1, len(testServer.MetricLines))
 	assert.Equal(t, "\"my-metric\" 20 source=\"localhost\"", testServer.MetricLines[0])
 }
+
+func TestEndToEndWithInternalMetrics(t *testing.T) {
+	waitTill1stInternalMetricsCollection := time.Millisecond * 1500
+	internalMetricsInterval := time.Millisecond * 900
+
+	testServer := startTestServer()
+	defer testServer.Close()
+
+	sender, err := NewSender(testServer.URL, InternalMetricsInterval(internalMetricsInterval))
+	require.NoError(t, err)
+	require.NoError(t, sender.SendMetric("my metric", 20, 0, "localhost", nil))
+	time.Sleep(waitTill1stInternalMetricsCollection)
```

Comment:
Thinking out loud:

We know we don't like having `time.Sleep` in our tests, but I think it's reasonable to do it in an integration test, since we're accepting some inefficiency in exchange for being able to test like a user does.

I think making the intervals fairly short here is a good approach. If the feedback loops are slow, we could try making the intervals even shorter, or we could expose a way of setting the `time.Ticker` itself directly, and I think that would allow us to make it tick imperatively in our tests.

Reply:
Thanks for looking it over! I found it challenging to find the shortest possible values for: `waitTill1stInternalMetricsCollection` and `internalMetricsInterval` to make this test pass predictably. 
- I wanted to leave enough time for the test server to spin up. 
- And I wanted `waitTill1stInternalMetricsCollection` to be long enough to get a first collection but not a second collection.

To reduce the intervals, I think I the next step could be to benchmark how long it takes for the test server to spin up.

Reply:
"expose a way of setting the `time.Ticker` itself directly -- Could you explain this a bit more?

Reply:
Sure!

So, right now, the test starts the server, sends some metrics, and waits for metrics flushing to happen in the background. We need to wait for the flushing to happen, but we also want the test to be fast. And like you said, we can only dial down the wait times so much before the tests start to become flaky.

One thought I had was that instead of:

```go
func SetInterval(interval time.Duration) RegistryOption {
	return func(registry *realMetricRegistry) {
		registry.reportTicker = time.NewTicker(interval)
	}
}
```

we could have

```go
func SetReportTicker(ticker *time.Ticker) RegistryOption {
	return func(registry *realMetricRegistry) {
		registry.reportTicker = ticker
	}
}
```

Then, instead of calling:
`SetInterval(1 * time.Second)`, we would have `SetReportTicker(time.newTicker(1 * time.Second))`.

So then maybe we could do something like this

```go
ticker := time.NewTicker(1 * time.Second)
SetReportTicker(&ticker)

// more setup

ticker.C <- time.Now()

// assert

```

The ticker works by sending messages to a channel, so if we hold the ticker, it looks like we can send messages to the channel too, which would allow us to trigger the metrics flushing deterministically without having to wait.

Reply:
Thank you for the clear and detailed explanation! 🙏 I'll try this out.

Reply:
I tried doing the refactor. The next step is to modify the test itself. I can pick that up tomorrow.

Reply:
When trying to send a message to the ticker's channel, I get an error `Invalid operation: internalMetricsTicker.C <- time.Now() (send to the receive-only type <- chan Time)`. 

I am interpreting this to mean that I can only receive/read messages on this channel (and not send/write messages). 

I wanted to put my thoughts out there for feedback, but I can keep digging.

Reply:
Yeah, exactly, darn!

Reply:
So, here's what I came up with, which is probably simpler than any of my previous suggestions.

I added a `Flush()` method to the `sdkmetrics.Registry` type, which for `realRegistry` just calls `report()`. That allows us to send the internal SDK metrics using just a method call.

The fake ticker could be made to work, but it's a lot of dependency injection to be doing in an integration test, so it's probably better not to go down that route.